modAionImpl/src/org/aion/zero/impl/db/AionBlockStore.java

```diff
@@ -241,6 +242,103 @@ public AionBlock getChainBlockByNumber(long number) {
         }
     }
 
+    /**
+     * Returns a range of main chain blocks.
+     *
+     * @param first the height of the first block in the requested range; this block must exist in
+     *     the blockchain and be above the genesis to return a non-null output
+     * @param last the height of the last block in the requested range; when requesting blocks in
+     *     ascending order the last element will be substituted with the best block if its height is
+     *     above the best known block
+     * @return a list containing consecutive main chain blocks with heights ranging according to the
+     *     given parameters; or {@code null} in case of errors or illegal request
+     * @apiNote The blocks must be added to the list in the order that they are requested. If {@code
+     *     first > last} the blocks are returned in descending order of their height, otherwise when
+     *     {@code first < last} the blocks are returned in ascending order of their height.
+     */
+    public List<AionBlock> getBlocksByRange(long first, long last) {
+        if (first <= 0L) {
+            return null;
+        }
+
+        lock.readLock().lock();
+
+        try {
+            AionBlock block = getChainBlockByNumber(first);
+            if (block == null) {
+                // invalid request
+                return null;
+            }
+
+            if (first == last) {
+                return List.of(block);
+            } else if (first > last) { // first is highest -> can query directly by parent hash
+                List<AionBlock> blocks = new ArrayList<>();
+                blocks.add(block);
+
+                for (long i = first - 1; i >= (last > 0 ? last : 1); i--) {
+                    block = getBlockByHash(block.getParentHash());
+                    if (block == null) {
+                        // the block should have been stored but null was returned above
+                        LOG.error(
+                                "Encountered a kernel database corruption: cannot find block at level {} in data store.",
+                                i);
+                        return null; // stops at any invalid data
```

Comment:
@AionJayT what do you think?

Reply:
If database corruption, should we recover it or shut down the kernel? Otherwise, I believe the node operator might not notice the error message.

Reply:
In the current implementation, this would be a corruption in the database. However, if we envision light client implementations it could be simply that the blocks are pruned or never imported.

Reply:
The issue here is that the kernel cannot fulfill a request. It's not a case where the missing functionality is encountered by a process that imports blocks. I cannot anticipate all the cases where the request made can be invalid, therefore, I don't think we should add more drastic measures here. If it's truly a db corruption it will be noted block imports as well.

Reply:
okay. I have no problem.